scala/scala-impl/src/org/jetbrains/plugins/scala/lang/refactoring/move/anonymousToInner/ScalaAnonymousToInnerHandler.scala

```diff
@@ -0,0 +1,157 @@
+  }
+
+  @VisibleForTesting
+  def parseInitialExtendsBlock(extendsBlock: ScExtendsBlock): (Array[ScalaVariableData], Either[ScFile, ScTemplateDefinition]) = {
+    val targetContainer = findTargetContainer(extendsBlock)
+    val usedVariables = collectUsedVariables(extendsBlock, targetContainer)
+    (usedVariables, targetContainer)
+  }
+
+  def performRefactoring(project: Project, className: String, variables: Array[ScalaVariableData], anonClass: ScExtendsBlock, originalElement: ScNewTemplateDefinition, targetContainer: Either[ScFile, ScTemplateDefinition]): Unit =
+    CommandProcessor.getInstance.executeCommand(project, () => {
+      val action: Runnable = () => {
+        try {
+          val innerClassNewTemplate = newTemplateForInnerClass(project, className, variables, originalElement)
+          val (newClassFromAnonymous, maybeWarningComment) = createClass(className, anonClass, variables, project)
+          val containerWithNewInnerClass = targetContainer
+            .fold(file => file, classOrObject => classOrObject)
+            .add(newClassFromAnonymous)
+
+          maybeWarningComment.foreach(comment => containerWithNewInnerClass.addBefore(comment, newClassFromAnonymous))
+          originalElement.replace(innerClassNewTemplate)
+        }
+        catch {
+          case e: IncorrectOperationException =>
+            LOG.error(e)
+        }
+
+      }
+      ApplicationManager.getApplication.runWriteAction(action)
+
+    }, getRefactoringName, null)
+
+  private def newTemplateForInnerClass(project: Project, name: String, variables: Array[ScalaVariableData], newTemplate: ScNewTemplateDefinition) = {
+    implicit val projectContext: ProjectContext = new ProjectContext(project)
+
+    val args = variables.map(_.variable.getName).mkString(", ")
+    val text = s"""new $name($args)""".stripMargin
+
+    createElementFromText[ScNewTemplateDefinition](text, newTemplate)
+  }
+
+  private def collectUsedVariables(anonClass: ScExtendsBlock, targetContainer: Either[ScFile, ScTemplateDefinition]): Array[ScalaVariableData] = {
+    var res: List[ScTypedDefinition] = Nil
+    anonClass.accept(new ScalaRecursiveElementVisitor() {
+      override def visitReferenceExpression(expression: ScReferenceExpression) = {
+        val refElement = expression.resolve
+        refElement match {
+          case p: ScParameter =>
+            val containingClass = p.getDeclarationScope
+            if (targetContainer.fold(_ => false, scTemplateDefinition => containingClass != scTemplateDefinition && !containingClass.isAncestorOf(scTemplateDefinition)))
+              res = p :: res
+          case r: ScReferencePattern =>
+            val containingClass = r.containingClass
+            if (targetContainer.fold(_ => false, scTemplateDefinition => containingClass != scTemplateDefinition && !containingClass.isAncestorOf(scTemplateDefinition) && !findTargetContainer(r).isLeft))
```

Comment:
This looks very similar to the code above.
Does this part matter
`!findTargetContainer(r).isLeft)`
?

If so, please, cover it with a separate test case.
It's not that obvious without context and it's easy to have regressions here.